taar/recommenders/similarity_recommender.py

```diff
@@ -0,0 +1,130 @@
+    used to rank specific addons for recommendation.
+
+    This will load a json file containing updated list of addon donors updated periodically
+    by a separate weekly process using Longitdudinal Telemetry data.
+
+    This recommender may provide useful recommendations when collaborative_recommender
+    may not work.
+    """
+
+    def __init__(self):
+        # Download the addon donors list.
+        self.donors_list = fetch_donors_list_remote_resource()
+
+        if self.donors_list is None:
+            logger.error("Cannot download the JSON resource: {}".format(DONOR_LIST_KEY))
+
+        # Download the probability mapping curves from similarity to likelihood of being a good donor.
+        self.lr_curves = fetch_lr_curve_remote_resource()
+
+        if self.lr_curves is None:
+            logger.error("Cannot download the JSON resource: {}".format(SIM_TO_P_FUN_KEY))
+
+    def can_recommend(self, client_data):
+        # We can't recommend if we don't have our data files.
+        if self.donors_list is None or self.lr_curves is None:
+            return False
+
+        # Check that the client info contains a non-None value for each required telemetry field.
+        required_fields = ["geoCity", "subsessionLength", "locale", "os", "bookmark_count", "tab_open_count",
+                           "total_uri", "unique_tlds"]
+
+        for single_required_field in required_fields:
+            client_value = client_data.get(single_required_field, None)
+            if client_value is None:
+                logger.error("Malformed client data encountered in SimilarityRecommender, {} field not detected"
+                             .format(single_required_field))
+                return False
+        return True
+
+    def get_lr(self, score):
+        # Compute the difference between each element in the lr curve score index and the current score.
+        index = [abs(score - s[0]) for s in self.lr_curves]
+
+        # Find the index of the closest value that was precomputed in lr_curves
+        _, idx = min((val, idx) for (idx, val) in enumerate(index))
+
+        numer_val = self.lr_curves[idx][1][0]
+        denum_val = self.lr_curves[idx][1][1]
+
+        # Compute LR based on numerator and denominator values
+        return float(numer_val) / float(denum_val)
+
+    def compute_similarity_to_donors(self, client_data):
+        # Define which features have only categorical (unicode) value possibilities and which are continuous.
+        cat_feats = ["geoCity", "locale", "os"]
+        continuous_feats = ["subsessionLength", "bookmark_count", "tab_open_count", "total_uri", "unique_tlds"]
+        client_categorical_feats = [client_data.get(specified_key) for specified_key in cat_feats]
+        client_continuous_feats = [client_data.get(specified_key) for specified_key in continuous_feats]
+
+        s = []
+        for donor in self.donors_list:
```

Comment:
Instead of writing our own loop here, can't we use something more performant like scipy's `cdist`? See [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html).

Reply:
Using cdist here, we would loose the ability to log an error with a specific donor. Cdist is indeed faster, but it is less felxible for custom metrics. I really prefer this loop for readability, diagnostics and debugging.